src/Microsoft.Extensions.ML/ServiceCollectionExtensions.cs

```diff
@@ -19,27 +20,50 @@ public static class ServiceCollectionExtensions
         /// The <see cref="IServiceCollection "/> to add services to.
         /// </param>
+        /// <param name="implementationFactory">
+        /// The factory that creates the <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/>.
+        /// </param>
         /// <returns>
         /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
         /// </returns>
         public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
-            this IServiceCollection services)
+            this IServiceCollection services,
+            Func<IServiceProvider, PredictionEnginePool<TData, TPrediction>> implementationFactory)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddPredictionEngineServices<TData, TPrediction>();
+            services
+                .AddPrerequisiteServices()
+                .AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>(implementationFactory);
             return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
         }
 
-        internal static IServiceCollection AddPredictionEngineServices<TData, TPrediction>(
+        /// <summary>
+        /// Adds a <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> to the service collection.
+        /// </summary>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection "/> to add services to.
+        /// </param>
+        /// <returns>
+        /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
+        /// </returns>
+        public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
             this IServiceCollection services)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddLogging();
-            services.AddOptions();
-            services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<MLOptions>, PostMLContextOptionsConfiguration>());
-            services.AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            services
+                .AddPrerequisiteServices()
+                .AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
+        }
+
+        public static IServiceCollection AddPrerequisiteServices(this IServiceCollection services)
+        {
+            services
+                .AddLogging()
+                .AddOptions()
+                .TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<MLOptions>, PostMLContextOptionsConfiguration>());
             return services;
```

Comment:
You could change this to simply:

```C#

return services
                .AddLogging()
                .AddOptions()
                .TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<MLOptions>, PostMLContextOptionsConfiguration>());
```

Reply:
Unfortunately, `TryAddEnumerable` returns void, so this isn't possible in this case.